File: bin/classes/com/RichardSzeto/checkers/Timer.cpp
/** Class that holds a timer
  *
  * @author Richard Szeto
  *
  *
  */

#include <iostream>
#include <ctime>
#include "Timer.h"

namespace CheckersSzeto
{
	Timer::Timer()
	{
		resetted = true;
		running = false;
		beg = 0;
		end = 0;
	}

	Timer::~Timer()
	{
		// deliberately left empty
	}

	void Timer::start()
	{
		if(!running)
		{
			if(resetted)
			{
				beg = static_cast<unsigned long>(clock());
			}
			else
			{
				beg = beg - end - static_cast<unsigned long>(clock());
			}

			running = true;

			resetted = true;
		}
	}

	void Timer::stop()
	{
		if(running)
		{
			end = static_cast<unsigned long>(clock());

			running = false;
		}
	}

	void Timer::reset()
	{
		bool wereRunning = running;

		if(wereRunning)
		{
			stop();
		}

		resetted = true;

		beg = 0;

		end = 0;

		if(wereRunning)
		{
			start();
		}
	}

	bool Timer::isRunning() const
	{
		return running;
	}

	unsigned long Timer::getTime() const
	{
		if(running)
		{
			return ((static_cast<unsigned long>(clock()) - beg) / static_cast<double>(CLOCKS_PER_SEC) * 1000);
		}
		else
		{
			return ((end - beg) / static_cast<double>(CLOCKS_PER_SEC) * 1000);
		}
	}

	bool Timer::isOver(const unsigned long &milliseconds) const
	{
		return milliseconds <= getTime();
	}
}

File: src/com/RichardSzeto/checkers/NotActivities/Timer.java
/** Class that holds a timer
  *
  * @author Richard Szeto
  *
  *
  */

package com.RichardSzeto.checkers.NotActivities;

public class Timer
{
    protected boolean resetted;
    protected boolean running;
    protected long beg;
    protected long end;
    
	public Timer()
	{
	    resetted = true;
        running = false;
        beg = 0;
        end = 0;
	}

	public void start()
	{
	    if(!running)
        {
            if(resetted)
            {
                beg = System.currentTimeMillis();
            }
            else
            {
                beg = beg - end - System.currentTimeMillis();
            }

            running = true;

            resetted = true;
        }
	}
	
	public void stop()
	{
	    if(running)
        {
            end = System.currentTimeMillis();

            running = false;
        }
	}
	
	public void reset()
	{
	    boolean wereRunning = running;

        if(wereRunning)
        {
            stop();
        }

        resetted = true;

        beg = 0;

        end = 0;

        if(wereRunning)
        {
            start();
        }
	}
	
	public boolean isRunning()
	{
	    return running;
	}
	
	public long getTime()
	{
	    if(running)
        {
            return System.currentTimeMillis() - beg;
        }
        else
        {
            return end - beg;
        }
	}
	
	public boolean isOver(long milliseconds)
	{
	    return milliseconds <= getTime();
	}
}
